src/web/model_api.nit

```diff
@@ -0,0 +1,124 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module model_api
+
+import web_base
+import highlight
+
+# Specific handler for nitweb API.
+abstract class APIHandler
+	super ModelHandler
+
+	# The JSON API does not filter anything by default.
+	#
+	# So we can cache the model view.
+	var view: ModelView is lazy do
+		var view = new ModelView(model)
+		view.min_visibility = private_visibility
+		view.include_fictive = true
+		view.include_empty_doc = true
+		view.include_attribute = true
+		view.include_test_suite = true
+		return view
+	end
+
+	# Try to load the mentity from uri with `/:id`.
+	#
+	# Send 400 if `:id` is null.
+	# Send 404 if no entity is found.
+	# Return null in both cases.
+	fun mentity_from_uri(req: HttpRequest, res: HttpResponse): nullable MEntity do
+		var id = req.param("id")
+		if id == null then
+			res.error 400
+			return null
+		end
+		var mentity = find_mentity(view, id)
+		if mentity == null then
+			res.error 404
+		end
+		return mentity
+	end
+end
+
+# Group all api handlers in one router.
+class APIRouter
+	super Router
+
+	# Model to pass to handlers.
+	var model: Model
+
+	# ModelBuilder to pass to handlers.
+	var modelbuilder: ModelBuilder
+
+	# Mainmodule to pass to handlers.
+	var mainmodule: MModule
+
+	init do
+		use("/list", new APIList(model, mainmodule))
+	end
+end
+# List all mentities.
+#
+# MEntities can be filtered on their kind using the `k` parameter.
+# Allowed kinds are `package`, `group`, `module`, `class`, `classdef`, `property`, `propdef`.
+#
+# List size can be limited with the `n` parameter.
+#
+# Example: `GET /list?k=module?n=10`
```

Comment:
If you want a `next` button or something, what about a `from` or `page` parameter?